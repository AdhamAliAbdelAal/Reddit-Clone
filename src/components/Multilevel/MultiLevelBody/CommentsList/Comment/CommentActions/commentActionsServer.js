// services
import axios from '../../../../../../services/instance';

export const saveCommenttoDB = (commentID, comment, setComment) => {
  console.log('saveCommentServer CommentID', commentID);
  console.log('oldComment', comment);
  axios.post(`/comments/${commentID}/save`).then((response) => {
    console.log('patch posts response to save comment ', response);
    // 200
    if (response.status === 200) {
      console.log('Saved Sucessfully :)');
      const oldComment = comment;
      // Toogle Save
      setComment(oldComment);
      alert('Comment Saved');
      console.log(response.data);
    } else if (response.status === 304) {
      alert('This Comment is already Saved');
    }
  }).catch((error) => {
    console.log(error);
    // 404
    // Error message:error.response.data.message
    alert(error?.response?.data?.errorMessage);
  });
};
