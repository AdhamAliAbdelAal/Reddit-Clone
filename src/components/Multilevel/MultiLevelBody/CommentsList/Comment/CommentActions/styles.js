import { Box } from '@mui/material';
import { styled } from '@mui/material/styles';

// CommentContainer
export const CommentActionsContainer = styled(Box)(() => ({
  width: '100%',
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',

  marginLeft: '-5px',

  '& > * :not(p)': {
    height: '100%',
    cursor: 'pointer',
  },

  // border: '2px solid black',
}));
