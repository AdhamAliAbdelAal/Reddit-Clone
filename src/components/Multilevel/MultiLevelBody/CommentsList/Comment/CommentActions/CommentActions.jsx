/* eslint-disable no-unused-vars */
import { useState } from 'react';
import { useCookies } from 'react-cookie';

// MUI Compoents
import ChatBubbleOutlineOutlinedIcon from '@mui/icons-material/ChatBubbleOutlineOutlined';
import MoreVertIcon from '@mui/icons-material/MoreVert';
import ShieldOutlinedIcon from '@mui/icons-material/ShieldOutlined';
import { Box } from '@mui/system';

// Components
import CommentReactions from './CommentReactions/CommentReactions';
import DropDownList from '../../../../../DropDownList/DropDownList';
import TextEditor from '../../../../../CreatePost/CreatePostContainer/CreatePostForm/TextEditor/TextEditor';
import RedditButton from '../../../../../RedditButton/RedditButton';

// Context
import { usePostContext } from '../../../../../../contexts/PostContext';

// styles
import { CommentActionsContainer } from './styles';
import { ElementBox, FooterText } from '../../../../../MainProfile/Posts/PostFooter/styles';
import { SaveButton } from '../../../MultiLevelPostContent/styles';

// Server
import { saveComment } from '../../commentsListServer';

function CommentActions(props) {
  // cookies
  // eslint-disable-next-line no-unused-vars
  const [cookies, setCookies] = useCookies(['redditUser']);

  // Context
  const { post } = usePostContext();
  const { comment } = props;

  // States
  const [reply, setReply] = useState('');
  const [replyEditor, setReplyEditor] = useState(false);

  const dropDownListOptions = [{ value: 'Save', icon: 'Save' }];
  const dropDownListOptionsAuthor = [{ value: 'Edit', icon: 'Edit' }, { value: 'Save', icon: 'Save' }, { value: 'Delete', icon: 'Delete' }];
  const dropDownListOptionsMod = [{ value: 'Approve', icon: 'Approve' }, { value: 'Remove', icon: 'Remove' }, { value: 'Remove as spam', icon: 'Remove as spam' }, { value: 'Lock comment', icon: 'Lock comment' }];

  const replyOnComment = () => {
    // call Reply endPoint
    console.log('Reply on Comment with Text', comment?.text);
    if (saveComment(comment?._id, 'Comment', reply)) {
      setReply('');
      setReplyEditor(false);

      // Need refresh post Component =>to pop comment
      // UpdatePost();
      // Refetch post
    }
  };

  const handleCommentReplyChange = (text) => {
    setReply(text);
  };

  const shareComment = () => {
    // call Share endPoint
    console.log('Share Comment with Text', comment?.text);
    // navigator.clipboard.writeText(this.state.textToCopy);
    navigator.clipboard.writeText('Basma');
  };
  const handleSelectEdit = (option) => {
    // Call Back API
    console.log('Option Selected', option);
    if (option === 'Edit') {
      // call Edit endPoint
      console.log('Edit Comment with Text', comment?.text);
    } else if (option === 'Save') {
      // call Save endPoint
      console.log('Save Comment with Text', comment?.text);
    } else if (option === 'Delete') {
      // call Delete endPoint
      console.log('Delete Comment with Text', comment?.text);
    }
  };

  const handleSelectMod = (option) => {
    console.log(option);
    // Call Back API
  };

  return (
    <div>

      <CommentActionsContainer>
        {comment && (
        <CommentReactions
          votes={comment?.votes}
          commentVoteStatus={comment?.votes}
          commentId={comment?._id}
        />
        )}
        <ElementBox onClick={() => setReplyEditor(true)}>
          <ChatBubbleOutlineOutlinedIcon />
          <FooterText variant="caption" responsiveshare={true.toString()}>
            Reply
          </FooterText>
        </ElementBox>

        <ElementBox onClick={shareComment}>
          <FooterText variant="caption" responsiveshare={false.toString()}>
            Share
          </FooterText>
        </ElementBox>

        <DropDownList handleSelectionFun={handleSelectEdit} options={cookies.redditUser?.id === comment?.author?._id ? dropDownListOptionsAuthor : dropDownListOptions}>
          <MoreVertIcon />
        </DropDownList>

        {/* Add cONDION ON THIS PREVIEW */}
        <DropDownList handleSelectionFun={handleSelectMod} options={dropDownListOptionsMod}>
          <ShieldOutlinedIcon />
        </DropDownList>

      </CommentActionsContainer>
      {replyEditor ? (
        <>
          <TextEditor
            handlePostTextChange={handleCommentReplyChange}
            postText={reply}
            commentPlaceholder
          />
          <Box m={2} gap={1} display="flex" justifyContent="flex-end">
            <RedditButton
              padding="3px 16px"
              fontSize={14}
              fontWeight="bold"
              type="submit"
              onClick={() => {
                setReplyEditor(false);
                setReply('');
              }}
            >
              Cancel
            </RedditButton>

            <SaveButton
              variant="contained"
              type="submit"
              onClick={replyOnComment}
              disabled={reply?.length === 8}
            >
              Reply
            </SaveButton>
          </Box>
        </>
      ) : null}

    </div>
  );
}

export default CommentActions;
