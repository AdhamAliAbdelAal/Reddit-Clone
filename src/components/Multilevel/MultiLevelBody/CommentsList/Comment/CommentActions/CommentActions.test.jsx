// 63a47f3710ebdf1da11a5b96
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import PostContextProvider from '../../../../../../contexts/PostContext';
import CommentActions from './CommentActions';

// test snapshot
test('CommentActions-test-snapshot', async () => {
  const tree = renderer.create(
    <PostContextProvider>
      <CommentActions />
    </PostContextProvider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
