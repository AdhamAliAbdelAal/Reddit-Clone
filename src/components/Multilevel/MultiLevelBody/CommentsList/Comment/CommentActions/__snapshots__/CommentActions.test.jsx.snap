// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentActions-test-snapshot 1`] = `
<div>
  <div
    className="MuiBox-root css-15u28o7"
  >
    <div
      className="MuiBox-root css-orlaaz"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ChatBubbleOutlineOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H6l-2 2V4h16v12z"
        />
      </svg>
      <span
        className="MuiTypography-root MuiTypography-caption css-g1q63z-MuiTypography-root"
        responsiveshare="true"
      >
        Reply
      </span>
    </div>
    <div
      className="MuiBox-root css-orlaaz"
      onClick={[Function]}
    >
      <span
        className="MuiTypography-root MuiTypography-caption css-1l4yqms-MuiTypography-root"
        responsiveshare="false"
      >
        Share
      </span>
    </div>
    <div>
      <button
        aria-haspopup="true"
        aria-label="more"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        id="long-button"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MoreVertIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
